### PR TITLE
Use assimp as mesh importer and exporter

### PR DESCRIPTION
**Pull request by servantf**
Use assimp to be able to import and export mesh instead of custom code.

Currently just replacing the obj import and export function, but maybe updated later to be more generic

### CODE REVIEW

File: src/aliceVision/mesh/Mesh.cpp [hunk 1]
```diff
@@ -2258,284 +2265,151 @@ void Mesh::getLargestConnectedComponentTrisIds(StaticVector<int>& out) const
...
-        long t1 = mvsUtils::initEstimate();
-        int idline = 0;
-        while(getline(in, line))
+    while (!nodes.empty())
+    {   
+        aiNode * node = nodes.back();
+        nodes.pop_back();
+
+        Eigen::Matrix4d T;
+        for (int i = 0; i < 4; i++)
```
Comment (simogasp) [inline src/aliceVision/mesh/Mesh.cpp:2310]:
```suggestion
        for (int i = 0; i < 4; ++i)
```

File: src/aliceVision/mesh/Mesh.cpp [hunk 2]
```diff
@@ -2258,284 +2265,151 @@ void Mesh::getLargestConnectedComponentTrisIds(StaticVector<int>& out) const
...
-                    );
-                }
-                else
-                {
-                    sscanf(line.c_str(), "v %lf %lf %lf", &pt.x, &pt.y, &pt.z);
-                }
-                pts.push_back(pt);
-            }
-            else if((line[0] == 'v') && (line[1] == 'n') && (line[2] == ' '))
+            for (int j = 0; j < 4; j++)
```
Comment (simogasp) [inline src/aliceVision/mesh/Mesh.cpp:2312]:
```suggestion
            for (int j = 0; j < 4; ++j)
```

File: src/aliceVision/mesh/Mesh.cpp [hunk 3]
```diff
@@ -2258,284 +2265,151 @@ void Mesh::getLargestConnectedComponentTrisIds(StaticVector<int>& out) const
...
-                // printf("%f %f %f\n", pt.x, pt.y, pt.z);
-                normals.push_back(pt);
+                T(i, j) = node->mTransformation[i][j];
             }
-            else if((line[0] == 'v') && (line[1] == 't') && (line[2] == ' '))
+        }
+
+        for (int idMesh = 0; idMesh < node->mNumMeshes; idMesh++)
+        {
+            unsigned int meshId = node->mMeshes[idMesh];
```
Comment (simogasp) [inline src/aliceVision/mesh/Mesh.cpp:2320]:
```suggestion
            const auto meshId = node->mMeshes[idMesh];
```

File: src/aliceVision/mesh/Mesh.cpp [hunk 4]
```diff
@@ -2258,284 +2265,151 @@ void Mesh::getLargestConnectedComponentTrisIds(StaticVector<int>& out) const
...
-                Voxel vertex, uvCoord, vertexNormal;
-                Voxel vertex2, uvCoord2, vertexNormal2;
-                bool ok = false;
-                bool withNormal = false;
-                bool withUV = false;
-                bool withQuad = false;
-                if(n2 == 0)
+                map_indices[idPoint] = pts.size();
+
+                aiVector3D v = mesh->mVertices[idPoint];
```
Comment (simogasp) [inline src/aliceVision/mesh/Mesh.cpp:2334]:
```suggestion
                const aiVector3D v = mesh->mVertices[idPoint];
```

File: src/aliceVision/mesh/Mesh.cpp [hunk 5]
```diff
@@ -2258,284 +2265,151 @@ void Mesh::getLargestConnectedComponentTrisIds(StaticVector<int>& out) const
...
+                pts.push_back(Point3d(v.x, v.y, v.z));
```
Comment (simogasp) [inline src/aliceVision/mesh/Mesh.cpp:2335]:
```suggestion
                pts.emplace_back(v.x, v.y, v.z);
```
requires to add the wrapper method to `StaticVector`

File: src/aliceVision/mesh/Mesh.cpp [hunk 6]
```diff
@@ -2258,284 +2265,151 @@ void Mesh::getLargestConnectedComponentTrisIds(StaticVector<int>& out) const
...
-                        vertexNormal2.y = vertexNormal.z;
-                        ok = true;
-                        withUV = true;
-                        withNormal = true;
-                        withQuad = true;
-                    }
+                    aiColor4D c = mesh->mColors[0][idPoint];
+                    double r = c.r * 255.0;
+                    double g = c.g * 255.0;
+                    double b = c.b * 255.0;
```
Comment (simogasp) [inline src/aliceVision/mesh/Mesh.cpp:2342]:
```suggestion
                    const double r = c.r * 255.0;
                    const double g = c.g * 255.0;
                    const double b = c.b * 255.0;
```

File: src/aliceVision/mesh/Mesh.cpp [hunk 7]
```diff
@@ -2258,284 +2265,151 @@ void Mesh::getLargestConnectedComponentTrisIds(StaticVector<int>& out) const
...
+                    _colors.push_back(aliceVision::rgb(r, g, b));
```
Comment (simogasp) [inline src/aliceVision/mesh/Mesh.cpp:2343]:
```suggestion
                    _colors.emplace_back(r, g, b);
```

File: src/aliceVision/mesh/Mesh.cpp [hunk 8]
```diff
@@ -2258,284 +2265,151 @@ void Mesh::getLargestConnectedComponentTrisIds(StaticVector<int>& out) const
...
-                               &vertex2.z, &vertexNormal2.z);
-                        vertex2.x = vertex.x; // same first point
-                        vertexNormal2.x = vertexNormal.x;
-                        vertex2.y = vertex.z; // 3rd point of the 1st triangle is the 2nd of the 2nd triangle.
-                        vertexNormal2.y = vertexNormal.z;
-                        ok = true;
-                        withNormal = true;
-                        withQuad = true;
-                    }
+                    aiVector3D t = mesh->mTextureCoords[0][idPoint];
```
Comment (simogasp) [inline src/aliceVision/mesh/Mesh.cpp:2348]:
```suggestion
                    const aiVector3D t = mesh->mTextureCoords[0][idPoint];
```

File: src/aliceVision/mesh/Mesh.cpp [hunk 9]
```diff
@@ -2258,284 +2265,151 @@ void Mesh::getLargestConnectedComponentTrisIds(StaticVector<int>& out) const
...
+                    uvCoords.push_back(Point2d(t.x, t.y));
```
Comment (simogasp) [inline src/aliceVision/mesh/Mesh.cpp:2349]:
```suggestion
                    uvCoords.emplace_back(t.x, t.y);
```
requires adding the method to `StaticVector`

File: src/aliceVision/mesh/Mesh.cpp [hunk 10]
```diff
@@ -2258,284 +2265,151 @@ void Mesh::getLargestConnectedComponentTrisIds(StaticVector<int>& out) const
...
                 }
-                if(!ok)
+                    
+                if (mesh->HasNormals())
                 {
-                    ALICEVISION_THROW_ERROR("Mesh: Unrecognized facet syntax while reading obj file: " << objAsciiFileName << ", line=" << idline << ", n1=" << n1 << ", n2=" << n2 << "\nline=\"" << line << "\"");
+                    aiVector3D n = mesh->mNormals[idPoint];
```
Comment (simogasp) [inline src/aliceVision/mesh/Mesh.cpp:2354]:
```suggestion
                    const aiVector3D n = mesh->mNormals[idPoint];
```

File: src/aliceVision/mesh/Mesh.cpp [hunk 11]
```diff
@@ -2258,284 +2265,151 @@ void Mesh::getLargestConnectedComponentTrisIds(StaticVector<int>& out) const
...
+                    normals.push_back(Point3d(n.x, n.y, n.z));
```
Comment (simogasp) [inline src/aliceVision/mesh/Mesh.cpp:2355]:
```suggestion
                    normals.emplace_back(n.x, n.y, n.z);      
```

File: src/aliceVision/mesh/Mesh.cpp [hunk 12]
```diff
@@ -2258,284 +2265,151 @@ void Mesh::getLargestConnectedComponentTrisIds(StaticVector<int>& out) const
...
+                    
+                if (mesh->HasNormals())
                 {
-                    ALICEVISION_THROW_ERROR("Mesh: Unrecognized facet syntax while reading obj file: " << objAsciiFileName << ", line=" << idline << ", n1=" << n1 << ", n2=" << n2 << "\nline=\"" << line << "\"");
+                    aiVector3D n = mesh->mNormals[idPoint];
+                    normals.push_back(Point3d(n.x, n.y, n.z));      
                 }
+            }
+
+            for (int idFace = 0; idFace < mesh->mNumFaces; idFace++)
```
Comment (simogasp) [inline src/aliceVision/mesh/Mesh.cpp:2359]:
```suggestion
            for (int idFace = 0; idFace < mesh->mNumFaces; ++idFace)
```

File: src/aliceVision/mesh/Mesh.cpp [hunk 13]
```diff
@@ -2258,284 +2265,151 @@ void Mesh::getLargestConnectedComponentTrisIds(StaticVector<int>& out) const
...
+            {
+                aiFace face = mesh->mFaces[idFace];
```
Comment (simogasp) [inline src/aliceVision/mesh/Mesh.cpp:2361]:
```suggestion
                const aiFace face = mesh->mFaces[idFace];
```

File: src/aliceVision/mesh/Mesh.cpp [hunk 14]
```diff
@@ -2258,284 +2265,151 @@ void Mesh::getLargestConnectedComponentTrisIds(StaticVector<int>& out) const
...
-                    if(withNormal)
-                    {
-                        trisNormalsIds.push_back(vertexNormal - Voxel(1, 1, 1));
-                    }
+                    continue;
                 }
 
-                // potential 2nd triangle
-                if(withQuad)
+                for (int idVertex = 0; idVertex < face.mNumIndices; idVertex++)
```
Comment (simogasp) [inline src/aliceVision/mesh/Mesh.cpp:2374]:
```suggestion
                for (int idVertex = 0; idVertex < face.mNumIndices; ++idVertex)
```

File: src/aliceVision/mesh/Mesh.cpp [hunk 15]
```diff
@@ -2258,284 +2265,151 @@ void Mesh::getLargestConnectedComponentTrisIds(StaticVector<int>& out) const
...
-                    triangle t;
-                    t.v[0] = vertex2.x - 1;
-                    t.v[1] = vertex2.y - 1;
-                    t.v[2] = vertex2.z - 1;
-                    t.alive = true;
-                    tris.push_back(t);
-                    _trisMtlIds.push_back(mtlId);
-                    if(withUV)
+                    unsigned int index = face.mIndices[idVertex];
+                    unsigned int global_index = map_indices[index];
```
Comment (simogasp) [inline src/aliceVision/mesh/Mesh.cpp:2377]:
```suggestion
                    const unsigned int index = face.mIndices[idVertex];
                    const unsigned int global_index = map_indices[index];
```

File: src/aliceVision/mesh/Mesh.cpp [hunk 16]
```diff
@@ -2258,284 +2265,151 @@ void Mesh::getLargestConnectedComponentTrisIds(StaticVector<int>& out) const
...
-                    mtlId = it->second;                 // already known material
+                    trisNormalsIds.push_back(nid);
                 }
-                maxMtlId = std::max(mtlId, maxMtlId);
             }
+        }
 
-            mvsUtils::printfEstimate(idline, nlines, t1);
-            idline++;
+        for (int idChild = 0; idChild < node->mNumChildren; idChild++)
```
Comment (simogasp) [inline src/aliceVision/mesh/Mesh.cpp:2404]:
```suggestion
        for (int idChild = 0; idChild < node->mNumChildren; ++idChild)
```

File: src/aliceVision/mesh/Texturing.cpp [hunk 17]
```diff
@@ -967,77 +970,112 @@ void Texturing::saveAsOBJ(const bfs::path& dir, const std::string& basename, ima
...
+        //Assimp does not allow vertex indices different from uv indices
+        //So we need to group and duplicate
+        std::map<std::pair<int, int>, int> unique_pairs;
         for(const auto triangleID : _atlases[atlasId])
         {
-            // vertex IDs
-            int vertexID1 = mesh->tris[triangleID].v[0];
-            int vertexID2 = mesh->tris[triangleID].v[1];
-            int vertexID3 = mesh->tris[triangleID].v[2];
+            for (int k = 0; k < 3; k++)
```
Comment (simogasp) [inline src/aliceVision/mesh/Texturing.cpp:1023]:
```suggestion
            for (int k = 0; k < 3; ++k)
```

File: src/aliceVision/mesh/Texturing.cpp [hunk 18]
```diff
@@ -967,77 +970,112 @@ void Texturing::saveAsOBJ(const bfs::path& dir, const std::string& basename, ima
...
+        aimesh->mNumVertices = unique_pairs.size();
+        aimesh->mVertices = new aiVector3D[unique_pairs.size()];
+        aimesh->mTextureCoords[0] = new aiVector3D[unique_pairs.size()];
 
-            fprintf(fobj, "f %i/%i %i/%i %i/%i\n", vertexID1 + 1, uvID1 + 1, vertexID2 + 1, uvID2 + 1, vertexID3 + 1, uvID3 + 1); // indexed from 1
+        int index = 0;
+        for (auto & p : unique_pairs)
+        {
+            int vertexId = p.first.first;
+            int uvId = p.first.second;
```
Comment (simogasp) [inline src/aliceVision/mesh/Texturing.cpp:1041]:
```suggestion
            const int vertexId = p.first.first;
            const int uvId = p.first.second;
```

File: src/aliceVision/mesh/Texturing.cpp [hunk 19]
```diff
@@ -967,77 +970,112 @@ void Texturing::saveAsOBJ(const bfs::path& dir, const std::string& basename, ima
...
+        aimesh->mFaces = new aiFace[aimesh->mNumFaces];
 
-    // header
-    fprintf(fmtl, "# \n");
-    fprintf(fmtl, "# Wavefront material file\n");
-    fprintf(fmtl, "# Created with AliceVision\n");
-    fprintf(fmtl, "# \n\n");
+        for(int i = 0; i < _atlases[atlasId].size(); i++)
+        {
+            int triangleId = _atlases[atlasId][i];
```
Comment (simogasp) [inline src/aliceVision/mesh/Texturing.cpp:1061]:
```suggestion
            const int triangleId = _atlases[atlasId][i];
```

File: src/aliceVision/mesh/Texturing.cpp [hunk 20]
```diff
@@ -967,77 +970,112 @@ void Texturing::saveAsOBJ(const bfs::path& dir, const std::string& basename, ima
...
-    FILE* fmtl = fopen(mtlFilename.c_str(), "w");
+        aimesh->mNumFaces = _atlases[atlasId].size();
+        aimesh->mFaces = new aiFace[aimesh->mNumFaces];
 
-    // header
-    fprintf(fmtl, "# \n");
-    fprintf(fmtl, "# Wavefront material file\n");
-    fprintf(fmtl, "# Created with AliceVision\n");
-    fprintf(fmtl, "# \n\n");
+        for(int i = 0; i < _atlases[atlasId].size(); i++)
```
Comment (simogasp) [inline src/aliceVision/mesh/Texturing.cpp:1059]:
```suggestion
        for(int i = 0; i < _atlases[atlasId].size(); ++i)
```

File: src/aliceVision/mesh/Texturing.cpp [hunk 21]
```diff
@@ -967,77 +970,112 @@ void Texturing::saveAsOBJ(const bfs::path& dir, const std::string& basename, ima
...
-        fprintf(fmtl, "Ns  0.0\n");
-        fprintf(fmtl, "illum 2\n");
-        fprintf(fmtl, "map_Kd %s\n", textureName.c_str());
+            aimesh->mFaces[i].mNumIndices = 3;
+            aimesh->mFaces[i].mIndices = new unsigned int[3];
+
+            for (int k = 0; k < 3; k++)
+            {
+                int vertexId = mesh->tris[triangleId].v[k];
+                int uvId = mesh->trisUvIds[triangleId].m[k];
```
Comment (simogasp) [inline src/aliceVision/mesh/Texturing.cpp:1069]:
```suggestion
                const int vertexId = mesh->tris[triangleId].v[k];
                const int uvId = mesh->trisUvIds[triangleId].m[k];
```

File: src/aliceVision/mesh/Texturing.cpp [hunk 22]
```diff
@@ -967,77 +970,112 @@ void Texturing::saveAsOBJ(const bfs::path& dir, const std::string& basename, ima
...
-        fprintf(fmtl, "Kd  0.6 0.6 0.6\n");
-        fprintf(fmtl, "Ks  0.0 0.0 0.0\n");
-        fprintf(fmtl, "d  1.0\n");
-        fprintf(fmtl, "Ns  0.0\n");
-        fprintf(fmtl, "illum 2\n");
-        fprintf(fmtl, "map_Kd %s\n", textureName.c_str());
+            aimesh->mFaces[i].mNumIndices = 3;
+            aimesh->mFaces[i].mIndices = new unsigned int[3];
+
+            for (int k = 0; k < 3; k++)
```
Comment (simogasp) [inline src/aliceVision/mesh/Texturing.cpp:1066]:
```suggestion
            for (int k = 0; k < 3; ++k)
```

**Review by simogasp — CHANGES_REQUESTED** [review]
Maybe add the cmake code to build assimp like the other libs
```
    set(ASSIMP_TARGET assimp)
    set(ASSIMP_BUILD_OPTIONS -DASSIMP_BUILD_ASSIMP_TOOLS:BOOL=OFF -DASSIMP_BUILD_TESTS:BOOL=OFF)
    set(ASSIMP_AV_VERSION 5.0.1)
    set(ASSIMP_FILENAME v${ASSIMP_AV_VERSION}.tar.gz)
    ExternalProject_Add(${ASSIMP_TARGET}
            URL https://github.com/assimp/assimp/archive/${ASSIMP_FILENAME}
            DOWNLOAD_NAME assimp-${ASSIMP_FILENAME}
            PREFIX ${BUILD_DIR}
            BUILD_IN_SOURCE 0
            BUILD_ALWAYS 0
            SOURCE_DIR ${CMAKE_CURRENT_BINARY_DIR}/assimp
            BINARY_DIR ${BUILD_DIR}/assimp_build
            INSTALL_DIR ${CMAKE_INSTALL_PREFIX}
            CONFIGURE_COMMAND ${CMAKE_COMMAND} ${CMAKE_CORE_BUILD_FLAGS} ${ASSIMP_BUILD_OPTIONS} -DCMAKE_INSTALL_PREFIX:PATH=<INSTALL_DIR> <SOURCE_DIR>
            )
    set(ASSIMP_CMAKE_FLAGS -DAssimp_DIR:PATH=${CMAKE_INSTALL_PREFIX}/lib/cmake/assimp-${ASSIMP_AV_VERSION}/)
```
Verify the last line as I cannot check if it still installs assimp in a folder named `assimp-${ASSIMP_VERSION}`

File: src/aliceVision/mesh/Texturing.cpp [hunk 23]
```diff
@@ -967,76 +970,112 @@ void Texturing::saveAsOBJ(const bfs::path& dir, const std::string& basename, ima
...
-    FILE* fobj = fopen(objFilename.c_str(), "w");
-
-    // header
-    fprintf(fobj, "# \n");
-    fprintf(fobj, "# Wavefront OBJ file\n");
-    fprintf(fobj, "# Created with AliceVision\n");
-    fprintf(fobj, "# \n");
-    fprintf(fobj, "mtllib %s\n\n", mtlName.c_str());
-    fprintf(fobj, "g TexturedMesh\n");
+    if (_atlases.size() == 0)
```
Comment (simogasp) [inline src/aliceVision/mesh/Texturing.cpp:973]:
```suggestion
    if (_atlases.empty)
```

**Comment by fabiencastan** [comment]
Is there a way to add a message in the OBJ file that it has been generated with AliceVision and specify the version?
Currently it says only assimp and the version of assimp.

**Review by fabiencastan — APPROVED** [review]

**Comment by simogasp** [comment]
> Is there a way to add a message in the OBJ file that it has been generated with AliceVision and specify the version?
> Currently it says only assimp and the version of assimp.

I think you can add it in the `aiMetadata`
https://assimp-docs.readthedocs.io/en/master/API/API-Documentation.html#_CPPv410aiMetadata

File: src/aliceVision/mesh/Mesh.cpp (hunk 5 above)
Comment (simogasp) [reply src/aliceVision/mesh/Mesh.cpp:2329]: I think adding this to `StaticVector` should work
```c++
    template< class... Args >
    void emplace_back( Args&&... args )
    {
        _data.emplace_back(std::forward<Args>(args)...);
    }
```

**Comment by fabiencastan** [comment]
I'm merging it for now.
@servantf Can you check if we can add metadata in a new PR?

**Comment by servantf** [comment]
There is no way to add metadata in obj files. 
aiMetadata are almost never exported but used for importers.
A solution may be to fork assimp and modify it ?

**Pull request merged**